### PR TITLE
Validate multihit branch lengths

### PR DESCRIPTION
**Pull request by willdumm**
See #123 for discussion, referencing a test in this PR

### CODE REVIEW

**Comment by willdumm** [comment]
The issues that ended up causing the discrepancy were:
* gradient-based optimization giving consistently incorrect (though close to correct on real data) results
* we were unnecessarily clamping linear-space probabilities to a number that was too large. Now using `torch.finfo(dtype).tiny` instead of `torch.finfo(dtype).eps`. The point is to avoid issues taking the log of something too close to zero, and the changed value is still adequate for that purpose.
* Also, there are two completely independent code paths computing codon probabilities, one for branch length optimization and one for loss computation (in DASM). This PR doesn't fix that, but tests both paths. A future PR should fix this.

File: netam/dxsm.py [hunk 1]
```diff
@@ -327,12 +327,20 @@ def _find_optimal_branch_length(
...
+        # print(aa_mask)
         sel_matrix = self.build_selection_matrix_from_parent_aa(
             aa_parents_indices, aa_mask
         )
+        # This is essential so that it is not interpreted as indices!!
+        assert aa_mask.dtype == torch.bool
         # Masks may be padded at end to account for sequences of different
         # lengths. The first part of the mask up to parent length should be
         # all the valid bits for the sequence.
+        # TODO how does this work for paired???
```
Comment (willdumm) [inline netam/dxsm.py:343]: I will resolve this before merge.

File: netam/molevol.py [hunk 2]
```diff
@@ -512,52 +536,181 @@ def log_pcp_probability(log_branch_length: torch.Tensor):
...
+        [
+            np.log(max(epsilon, np.exp(suggested_middle) - 0.01)),
+            np.log(np.exp(suggested_middle) + 0.01),
+        ],
+        [suggested_middle - 0.5, suggested_middle + 0.5],
+        [suggested_middle - 1.0, suggested_middle],
+        [suggested_middle + 1.0, suggested_middle],
+        [suggested_middle - 2.0, suggested_middle],
+        [suggested_middle + 2.0, suggested_middle],
+        [(end - start) / 2, suggested_middle],
```
Comment (willdumm) [inline netam/molevol.py:579]: This is absurd, and in my tests the first option works for all 50k pcps I tested. However, the first option was the result of much trial and error including the other options on this list, so I leave them all to hopefully handle future insane, unforeseen  PCPs and suggested optima.

File: tests/test_sequences.py [hunk 3]
```diff
@@ -198,3 +201,31 @@ def test_dataset_inputs_of_pcp_df(pcp_df, pcp_df_paired):
...
+    test_nt_codon_idxs = unflatten_codon_idxs(true_codon_indices)
+    assert torch.allclose(test_nt_codon_idxs, true_nt_codon_indices)
+
+    test_codon_idxs = flatten_codon_idxs(true_nt_codon_indices)
+
+    print(test_codon_idxs)
+    print(true_codon_indices)
+    if not torch.allclose(test_codon_idxs, true_codon_indices):
+        print(test_codon_idxs)
+        print(true_codon_indices)
```
[the remark this answers is not in the document]
Comment (willdumm) [reply tests/test_sequences.py:230]: When tests fail, this is exactly the info I need to see. They're hidden when tests pass. I'll leave them.

File: netam/molevol.py [hunk 4]
```diff
@@ -512,52 +536,181 @@ def log_pcp_probability(log_branch_length: torch.Tensor):
...
+        brute_linear_step: Step size for brute force search.
 
-        optimizer.zero_grad()
+    Returns:
+        A tuple containing the start, middle, and end points of the bracket.
+        start and end may be outside the range of the passed start and end, but middle point will be contained in the passed range.
+    """
+    if suggested_middle <= start:
+        suggested_middle = np.log(np.exp(start) + 2 * linear_step)
+    if suggested_middle >= end:
```
[the remark this answers is not in the document]
Comment (willdumm) [reply netam/molevol.py]: fixed

**Review by matsen — APPROVED** [review]
Merge when happy
